org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/FinallyFilter.java

```diff
@@ -0,0 +1,196 @@
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+/**
+ * Filters duplicates of finally blocks that compiler generates.
+ */
+public final class FinallyFilter implements IFilter {
+
+	public void filter(final String className, final String superClassName,
+			final MethodNode methodNode, final IFilterOutput output) {
+		for (final TryCatchBlockNode tryCatchBlock : methodNode.tryCatchBlocks) {
+			if (tryCatchBlock.type == null) {
+				new Matcher(output, methodNode, tryCatchBlock).filter();
+			}
+		}
+	}
+
+	private static class Matcher {
+		final IFilterOutput output;
+		final MethodNode m;
+		final TryCatchBlockNode finallyBlock;
+
+		/**
+		 * Instructions covered by finally handler.
+		 */
+		final Set<AbstractInsnNode> inside = new HashSet<AbstractInsnNode>();
+
+		private Matcher(final IFilterOutput output, final MethodNode m,
+				final TryCatchBlockNode finallyBlock) {
+			this.output = output;
+			this.m = m;
+			this.finallyBlock = finallyBlock;
+		}
+
+		private void filter() {
+			final AbstractInsnNode e = next(finallyBlock.handler);
+			final int size = size(e);
+			if (size <= 0) {
+				return;
+			}
+
+			for (final TryCatchBlockNode t : m.tryCatchBlocks) {
+				if (t.handler == finallyBlock.handler) {
+					AbstractInsnNode i = t.start;
+					while (i != t.end) {
+						inside.add(i);
+						i = i.getNext();
+					}
+				}
+			}
+
+			for (final TryCatchBlockNode t : m.tryCatchBlocks) {
+				if (t.handler == finallyBlock.handler) {
+					boolean continues = false;
+					AbstractInsnNode i = t.start;
+
+					while (i != t.end) {
+						switch (i.getType()) {
+						case AbstractInsnNode.FRAME:
+						case AbstractInsnNode.LINE:
+						case AbstractInsnNode.LABEL:
+							break;
+						case AbstractInsnNode.JUMP_INSN:
+							final AbstractInsnNode jumpTarget = next(
+									((JumpInsnNode) i).label);
+							if (!inside.contains(jumpTarget)) {
+								merge(size, e, jumpTarget);
+							}
+							continues = i.getOpcode() != Opcodes.GOTO;
+							break;
+						default:
+							switch (i.getOpcode()) {
+								case Opcodes.IRETURN:
+								case Opcodes.LRETURN:
+								case Opcodes.FRETURN:
+								case Opcodes.DRETURN:
+								case Opcodes.ARETURN:
+								case Opcodes.RETURN:
+								case Opcodes.ATHROW:
+									continues = false;
+									break;
+								default:
+									continues = true;
+									break;
+							}
+							break;
+						}
+						i = i.getNext();
+					}
+
+					i = next(i);
+					if (continues && !inside.contains(i)) {
+						merge(size, e, i);
+					}
+				}
+
+				if (t != finallyBlock && t.start == finallyBlock.start
+						&& t.end == finallyBlock.end) {
+					final AbstractInsnNode i = next(next(t.handler));
```

Comment:
@Godin Potential NPE? `next(t.handler)` could possibly return `null`

Reply:
@marchof `t.handler` is a label that must have instruction after it, so `next(t.handler)` can't be `null`.

`next(next(t.handler))` potentially can be `null` only if handler has exactly one instruction that is the last one in method, but in this case safe to call `merge(size, e, null)` that safely calls `isSame(size, e, null)` and returns.